### PR TITLE
LibWeb: Parse the keyword 'none' in the CSS4 color functions

### PR DESCRIPTION
**Pull request by qligier**
This PR updates the CSS parser to support the keyword 'none' in the CSS4 color functions. The underlying CSSColorValue already supports this keyword, meaning the parser can instantiate the color directly.

This removes 3 TODOs and fixes a few WPT tests, mainly in `css/css-color/parsing` and in HSL tests.
In the new tests, I avoided colors that have rounding issues, so these should be stable in time.

### CODE REVIEW

File: Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp [hunk 1]
```diff
@@ -2583,19 +2583,22 @@ RefPtr<CSSStyleValue> Parser::parse_number_value(TokenStream<ComponentValue>& to
...
-    }
+    // Parses [<percentage> | <number>] (which is equivalent to [<alpha-value>])
+    if (auto number = parse_number_value(tokens))
+        return number;
+    if (auto percentage = parse_percentage_value(tokens))
+        return percentage;
+
+    return nullptr;
+}
+
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:2594]: I see why you wanted to shrink this methods, but it was done the previous way to avoid parsing `calc()` multiple times. If we have `calc(10%)` for example, `parse_number_percentage_value()` will now:
1. `parse_number_value()`
	1. Try to parse a number; fail
	2. Try to parse a `calc()`
	3. See that the `calc()` doesn't resolve to a number; fail
2. `parse_percentage_value()`
	1. Try to parse a percentage; fail
	2. Try to parse a `calc()`, again, doing the same work as in step 1.ii
	3. See that the `calc()` resolves to percentage and return it.

File: Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp [hunk 2]
```diff
@@ -2583,19 +2583,22 @@ RefPtr<CSSStyleValue> Parser::parse_number_value(TokenStream<ComponentValue>& to
...
+RefPtr<CSSStyleValue> Parser::parse_number_percentage_none_value(TokenStream<ComponentValue>& tokens)
+{
+    // Parses [<percentage> | <number> | none] (which is equivalent to [<alpha-value> | none])
+    if (auto value = parse_number_percentage_value(tokens))
+        return value;
+    if (auto keyword = parse_keyword_value(tokens); keyword && keyword->as_keyword().keyword() == Keyword::None)
+        return keyword;
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:2601]: If there's a keyword here that isn't `none`, the token has still been consumed. You need to wrap this in a `TokenStream` transaction, and then only commit it if the keyword was `none`.

File: Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp [hunk 3]
```diff
@@ -2841,29 +2844,33 @@ RefPtr<CSSStyleValue> Parser::parse_rect_value(TokenStream<ComponentValue>& toke
...
+RefPtr<CSSStyleValue> Parser::parse_hue_none_value(TokenStream<ComponentValue>& tokens)
 {
-    // <hue> = <number> | <angle>
+    // Parses [<hue> | none]
+    //   <hue> = <number> | <angle>
     if (auto number = parse_number_value(tokens))
         return number;
     if (auto angle = parse_angle_value(tokens))
         return angle;
+    if (auto keyword = parse_keyword_value(tokens); keyword && keyword->as_keyword().keyword() == Keyword::None)
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:2855]: Same point about consuming tokens and using a transaction.

File: Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp [hunk 4]
```diff
@@ -3056,6 +3070,8 @@ RefPtr<CSSStyleValue> Parser::parse_hsl_color_value(TokenStream<ComponentValue>&
             inner_tokens.discard_whitespace();
 
             alpha = parse_number_percentage_value(inner_tokens);
+            if (!alpha)
+                return {};
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:3074]: I got a bit confused by this at first, but nice catch! Could you move this fix into its own commit? It's unrelated to the other changes, and that would let you add a commit description for it.

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]
Hi! Thanks for working on this. Generally good (and thanks for the tests!) but I spotted a couple of issues.

File: Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp (hunk 4 above)
Comment (qligier) [reply Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:3074]: Sure, I'll also add a comment then!

File: Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp (hunk 1 above)
Comment (qligier) [reply Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:2594]: Indeed, there's an easy optimization here.

File: Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp (hunk 2 above)
Comment (qligier) [reply Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:2601]: I hadn't thought of that, of course!

File: Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp (hunk 4 above)
Comment (qligier) [reply Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:2920]: Moved to another commit!

File: Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp (hunk 2 above)
Comment (qligier) [reply Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp]: Fixed

File: Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp (hunk 3 above)
Comment (qligier) [reply Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp]: Fixed

**Comment by qligier** [comment]
Fixed the issues mentioned in feedback and rebased on master.

**Review by AtkinsSJ — APPROVED** [review]
This all looks good to me!
